### PR TITLE
[SPARK] Add benchmark for Spark TRowSet generation of row-based and column-based

### PR DESCRIPTION
**Pull request by bowenliang123**
# :mag: Description
## Issue References 🔗

Subtask of #5808.

## Describe Your Solution 🔧

Add performance benchmark for Spark TRowSet generation for
-  row-based TRowSet on HIVE_CLI_SERVICE_PROTOCOL_V5 and below
-  column-based TRowSet on HIVE_CLI_SERVICE_PROTOCOL_V6 and above

## Types of changes :bookmark:

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

#### Behavior With This Pull Request :tada:
Row-based:
image

Column-based:
image

#### Related Unit Tests
Added "to row set benchmark" ut in Spark Engine's RowSetSuite.

---

# Checklists
## 📝 Author Self Checklist

- [x] My code follows the [style guidelines](https://kyuubi.readthedocs.io/en/master/contributing/code/style.html) of this project
- [x] I have performed a self-review
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] This patch was not authored or co-authored using [Generative Tooling](https://www.apache.org/legal/generative-tooling.html)

## 📝 Committer Pre-Merge Checklist

- [ ] Pull request title is okay.
- [ ] No license issues.
- [ ] Milestone correctly set?
- [ ] Test coverage is ok
- [ ] Assignees are selected.
- [ ] Minimum number of approvals
- [ ] No changes are requested

**Be nice. Be informative.**

### CODE REVIEW

**Comment by bowenliang123** [comment]
> Should we create a new class separately? And you can refer to `org.apache.spark.sql.ZorderCoreBenchmark`.

Thanks for the advice. Moved to a new class `RowSetBenchmark`.

File: externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/schema/RowSetBenchmark.scala [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+import scala.math.BigDecimal.RoundingMode
+
+import org.apache.commons.lang3.time.StopWatch
+import org.apache.hive.service.rpc.thrift.TProtocolVersion
+import org.apache.spark.sql.Row
+import org.apache.spark.sql.types._
+
+class RowSetBenchmark extends BaseRowSetSuite {
+  test("to row set benchmark") {
```
Comment (yaooqinn) [inline externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/schema/RowSetBenchmark.scala:28]: don't put benchmarks into tests

Comment (bowenliang123) [reply externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/schema/RowSetBenchmark.scala]: Would like to keep it as running via tests by setting `RUN_BENCHMARK=1`, just like other existed benchmarks like TPCDSTableGenerateBenchmark.

JMH for isolated benchmark testing could be introduced next time.
Having trouble in integrating JMH for Scala without official Maven plugin support , using JMH Java annotations , the  proper execution entry point to run with JMH and the isolation path for JMH benchmarks.

Comment (bowenliang123) [reply externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/schema/RowSetBenchmark.scala]: Hi @yaooqinn . I have changed to use Spark's Benchmark tools for running and generating the benchmarks, just like the `TPCDSTableGenerateBenchmark` and `ZorderCoreBenchmark`. Could you have a look? 
And thanks for the guidance from @wForget .

Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/schema/RowSetBenchmark.scala]: -1 to use Spark's Benchmark tools in engine modules

Comment (bowenliang123) [reply externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/schema/RowSetBenchmark.scala]: We could refactor this onto JMH in the follow-up PRs. These tests are not run with GA tests. This should not be a blocker issue here for evaluating the overall TRowSet generation.

File: externals/kyuubi-spark-sql-engine/pom.xml [hunk 2]
```diff
@@ -72,6 +72,13 @@
             <scope>test</scope>
         </dependency>
 
+        <dependency>
+            <groupId>org.apache.spark</groupId>
+            <artifactId>spark-core_${scala.binary.version}</artifactId>
+            <type>test-jar</type>
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/pom.xml:78]: we can get rid of this dependencies if we fork `org.apache.spark.benchmark.Benchmark` from Spark

File: externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/schema/RowSetBenchmark.scala (hunk 1 above)
Comment (pan3793) [reply externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/schema/RowSetBenchmark.scala]: > -1 to use Spark's Benchmark tools in engine modules

what's the reason/major concern

File: externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/spark/kyuubi/TRowSetBenchmark.scala [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+import org.apache.kyuubi.KyuubiFunSuite
+import org.apache.kyuubi.engine.spark.schema.{RowSet, RowSetHelper}
+
+/**
+ * Benchmark to measure the performance of generate TRowSet.
+ *
+ * {{{
+ *   RUN_BENCHMARK=1 ./build/mvn clean test \
+ *   -pl externals/kyuubi-spark-sql-engine -am \
+ *   -Dtest=none -DwildcardSuites=oorg.apache.spark.kyuubi.TRowSetBenchmark
```
Comment (pan3793) [inline externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/spark/kyuubi/TRowSetBenchmark.scala:38]:
```suggestion
 *   -Dtest=none -DwildcardSuites=org.apache.spark.kyuubi.TRowSetBenchmark
```

File: externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/schema/RowSetBenchmark.scala (hunk 1 above)
Comment (yaooqinn) [reply externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/schema/RowSetBenchmark.scala]: > We could refactor this onto JMH in the follow-up PRs

We don't need to refactor if it's originally designed with JMH

**Comment by yaooqinn** [comment]
And based on your screenshots in your PR desc, what are actually the control group, experimental groups？

File: externals/kyuubi-spark-sql-engine/pom.xml (hunk 2 above)
Comment (bowenliang123) [reply externals/kyuubi-spark-sql-engine/pom.xml:78]: The test-jar of `spark-core` must be included, as the `org.apache.spark.benchmark.Benchmark` class is part of it. The tests cannot be run without it.

**Comment by bowenliang123** [comment]
> And based on your screenshots in your PR desc, what are actually the control group, experimental groups？

There is no control or experimental group in this PR. It provides a benchmark tool for evaluating both column-based and row-based rowset. In the coming-up experiments, the benchmark will be run on the base version and different improvement implementations for comparison.

**Review by pan3793 — APPROVED** [review]
LGTM, I agree to adopt this benchmark framework, because

1. it is light 
2. there are existing benchmarks based on it
3. many Kyuubi developers are familiar with it

**Comment by bowenliang123** [comment]
And we could decouple it with Spark's utils and move it to kyuubi-util module for a general light-weight benchmark kit in the future. And when it's ready to integrate JMH in Kyuubi with Maven + sbt + Scala, this benchmark toolkit is able to be removed.

**Review by yaooqinn — CHANGES_REQUESTED** [review]
I will keep my -1 as the testing purpose here is not clear

**Comment by yaooqinn** [comment]
>  In the coming-up experiments

If there are a bunch of PRs, I suggest you create an umbrella, and an KPIP(discuss/vote in the dev list) is necessary for introduce a benchmarking framework.

**Comment by yaooqinn** [comment]
@wForget 

You are correct about the purpose of this PR, but the benchmark itself needs to be corrected. Technically, if we introduce the Spark benchmark tool, the first line of results in each single benchmark should be the control group as it always produces 1x for `Relative`.

The current test also varies the simple rule of `univariate analysis` for experiments. What I saw in the results is a mess, TBH.

**Comment by wForget** [comment]
> You are correct about the purpose of this PR, but the benchmark itself needs to be corrected. Technically, if we introduce the Spark benchmark tool, the first line of results in each single benchmark should be the control group as it always produces 1x for `Relative`.
> 
> The current test also varies the simple rule of `univariate analysis` for experiments. What I saw in the results is a mess, TBH.

Got it, thank you for your explanation.

**Comment by yaooqinn** [comment]
> I'm strongly against your comment here. ...  I respect all your comments 

Hi @bowenliang123. First thing first, calm down.

I want to clarify that I am a regular contributor/PMC member of Apache Kyuubi, just like everyone else. My comments on this PR are simply my personal opinion. I have left a veto with explanations, which also have been challenged and discussed.


> I did no evil and did not violate any community code of conduct now and ever! WHY make it difficult for me !!!

I know you well in person. You and nobody else don't violate CoC in this PR.

> Third, you told me to raise a KPIP for such a duplicated framework from a copied implementation.

`When in doubt, if a committer thinks a change needs a KPIP, it does.` --- https://kyuubi.apache.org/improvement-proposals.html

File: externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/schema/RowSetBenchmark.scala (hunk 1 above)
Comment (bowenliang123) [reply externals/kyuubi-spark-sql-engine/src/test/scala/org/apache/kyuubi/engine/spark/schema/RowSetBenchmark.scala]: [jmh-scala-benchmark-archetype](https://mvnrepository.com/artifact/org.openjdk.jmh/jmh-scala-benchmark-archetype)

**Pull request closed without merging**